File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/Board.java
package org.silnith.game.solitaire;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import org.silnith.deck.Card;
import org.silnith.deck.Suit;
import org.silnith.util.Pair;


/**
 * The board for Klondike solitaire.
 * 
 * <p>The board consists of seven columns, the foundation, and the stock pile.</p>
 */
public class Board {
    
    private final List<Column> columns;
    
    private final List<Card> stockPile;
    
    private final int stockPileIndex;
    
    private final Map<Suit, List<Card>> foundation;
    
    /**
     * Creates a new board.
     * 
     * <p>Parameters should be immutable.  This is not enforced in code.</p>
     * 
     * @param columns the columns for the new board
     * @param stockPile the stock pile
     * @param stockPileIndex the index into the stock pile of the current draw card.
     *         {@code 0} means no card is available to be drawn, {@code stockPile.size()}
     *         means all cards have been advanced and the last card is available to be drawn.
     * @param foundation the foundation
     */
    public Board(final List<Column> columns, final List<Card> stockPile, final int stockPileIndex,
            final Map<Suit, List<Card>> foundation) {
        super();
        if (stockPileIndex < 0) {
        	throw new IllegalArgumentException("Stock pile index must be non-negative.");
        }
        if (stockPileIndex > stockPile.size()) {
            throw new IllegalArgumentException("Stock pile index outside of stock pile.");
        }
        this.columns = columns;
        this.stockPile = stockPile;
        this.stockPileIndex = stockPileIndex;
        this.foundation = foundation;
    }
    
    /**
     * Returns the columns.
     * 
     * @return the columns
     */
    public List<Column> getColumns() {
        return columns;
    }
    
    /**
     * Returns the given column.
     * 
     * @param index the index of the column
     * @return the column
     * @throws IndexOutOfBoundsException if the index is out of bounds
     */
    public Column getColumn(int index) {
        return columns.get(index);
    }
    
    /**
     * Returns the stock pile.
     * 
     * @return the stock pile
     */
    public List<Card> getStockPile() {
        return stockPile;
    }
    
    /**
     * Returns the current index into the stock pile.
     * This is zero if the stock pile has not been advanced,
     * meaning no cards are available to be drawn.
     * When this reaches the size of the stock pile, it is
     * eligible to be recycled.
     * 
     * @return the current index into the stock pile.
     *         This is the number of cards that have been
     *         advanced.
     */
    public int getStockPileIndex() {
        return stockPileIndex;
    }
    
    /**
     * Returns the foundation.  The game is won when all
     * cards have been moved to the foundation.
     * 
     * @return the foundation
     */
    public Map<Suit, List<Card>> getFoundation() {
        return foundation;
    }
    
    /**
     * Returns {@code true} if the stock pile can be advanced.
     * If the stock pile is empty, or if the stock pile index
     * is beyond the end of the stock pile, this is {@code false}.
     * 
     * <p>If the stock pile is empty, it cannot be advanced or recycled.
     * Otherwise, either this or {@link #canRecycleStockPile()}
     * will be {@code true}, and the other {@code false}.</p>
     * 
     * @return {@code true} if the stock pile can be advanced
     */
    public boolean canAdvanceStockPile() {
        return stockPileIndex < stockPile.size();
    }
    
    /**
     * Returns {@code true} if the stock pile can be recycled.
     * If the stock pile is empty, or if the stock pile index
     * is not currently beyond the end of the stock pile, this is {@code false}.
     * 
     * <p>If the stock pile is empty, it cannot be advanced or recycled.
     * Otherwise, either this or {@link #canAdvanceStockPile()}
     * will be {@code true}, and the other {@code false}.</p>
     * 
     * @return {@code true} if the stock pile can be recycled
     */
    public boolean canRecycleStockPile() {
        // 0 represents no cards flipped, so offset by one
        return !stockPile.isEmpty() && stockPileIndex >= stockPile.size();
    }
    
    /**
     * Returns the current card that can be drawn from the stock pile.
     * 
     * @return the current card available from the stock pile
     * @throws IndexOutOfBoundsException if no card is available
     *         to be drawn from the stock pile
     */
    public Card getStockPileCard() {
        return stockPile.get(stockPileIndex - 1);
    }
    
    /**
     * Returns the top card on the foundation for the given suit.
     * 
     * @param suit the suit
     * @return the current top card for the suit in the foundation
     * @throws IndexOutOfBoundsException if the foundation has no cards for the given suit
     */
    public Card getTopOfFoundation(final Suit suit) {
        final List<Card> foundationForSuit = foundation.get(suit);
        return foundationForSuit.get(foundationForSuit.size() - 1);
    }
    
    /**
     * Returns a copy of the foundation with the given card added to it.
     * 
     * <p>This does no validation that the move is legal.</p>
     * 
     * @param card the card to add to the foundation
     * @return a copy of the foundation with one card added
     */
    public Map<Suit, List<Card>> getFoundationPlusCard(final Card card) {
        final Map<Suit, List<Card>> newFoundation = new EnumMap<>(foundation);
        final Suit suit = card.getSuit();
        final List<Card> stackForSuit = newFoundation.get(suit);
        final List<Card> newStackForSuit = new ArrayList<>(13);
        newStackForSuit.addAll(stackForSuit);
        newStackForSuit.add(card);
        newFoundation.put(suit, newStackForSuit);
        
        return newFoundation;
    }
    
    /**
     * Extracts the top card from the foundation for the given suit, and returns
     * both the card and the remaining foundation missing the card.
     * 
     * @param suit the suit of the foundation from which to remove the top card
     * @return a pair of the card and the remaining foundation
     */
    public Pair<Card, Map<Suit, List<Card>>> extractCardFromFoundation(final Suit suit) {
    	final List<Card> foundationForSuit = foundation.get(suit);
		final int suitCountMinusOne = foundationForSuit.size() - 1;
		final Card card = foundationForSuit.get(suitCountMinusOne);
		final Map<Suit, List<Card>> newFoundation = new EnumMap<>(foundation);
		final List<Card> newStackForSuit = new ArrayList<>(foundationForSuit.subList(0, suitCountMinusOne));
		newFoundation.put(suit, newStackForSuit);
		return new Pair<>(card, newFoundation);
    }
    
    /**
     * Extracts the currently visible card from the stock pile, and returns
     * both the card and the remaining stock pile missing the card.
     * 
     * @return a pair of the card, and the remaining stock pile
     * @throws IndexOutOfBoundsException if no card is available
     *         to be drawn from the stock pile
     */
    public Pair<Card, List<Card>> extractStockPileCard() {
    	final Card card = getStockPileCard();
		/*
		 * StockPile: [a, b, c]
		 * StockPileIndex: 2
		 * 
		 * Result: [a, c]
		 */
		final int size = stockPile.size();
		final List<Card> newStockPile = new ArrayList<>(size - 1);
		newStockPile.addAll(stockPile.subList(0, stockPileIndex - 1));
		newStockPile.addAll(stockPile.subList(stockPileIndex, size));
		return new Pair<Card, List<Card>>(card, newStockPile);
    }
    
    /**
     * Returns whether it would be legal to add the given card to the foundation.
     * 
     * @param card the card to check
     * @return {@code true} if it is legal to add the card to the foundation
     */
    public boolean canAddToFoundation(final Card card) {
    	return card.getValue().getValue() == 1 + foundation.get(card.getSuit()).size();
    }
    
    private void printCardTo(final PrintStream out, final Card card) {
    	out.format(Locale.US, "%2s%s", card.getValue().toSymbol(), card.getSuit().toSymbol());
    }
    
    /**
     * Prints an ASCII art representation of the board to the given output stream.
     * 
     * @param out the output stream
     */
    public void printTo(final PrintStream out) {
    	for (final Map.Entry<Suit, List<Card>> entry : foundation.entrySet()) {
    		out.format(Locale.US, "%2s", "");
    		final List<Card> cards = entry.getValue();
    		if (cards.isEmpty()) {
    			out.format(Locale.US, "%3s", "--");
    		} else {
    			printCardTo(out, cards.get(cards.size() - 1));
    		}
    	}
    	out.format(Locale.US, "%3s", "");
    	out.format(Locale.US, "(%2d/%2d)", stockPileIndex, stockPile.size());
    	out.format(Locale.US, "%2s", "");
    	if (stockPileIndex > 0) {
    		printCardTo(out, stockPile.get(stockPileIndex - 1));
    	} else {
    		out.format(Locale.US, "%3s", "");
    	}
    	out.println();
    	final List<Iterator<Card>> iterators = new ArrayList<Iterator<Card>>(columns.size());
    	for (final Column column : columns) {
        	out.format(Locale.US, "%2s", "");
        	out.format(Locale.US, "(%d)", column.getNumberOfFaceDownCards());
    		iterators.add(column.getFaceUpCards().iterator());
    	}
    	out.println();
    	boolean printedSomething;
    	do {
    		printedSomething = false;
    		for (final Iterator<Card> iterator : iterators) {
    			out.format(Locale.US, "%2s", "");
    			if (iterator.hasNext()) {
    				final Card card = iterator.next();
    				printCardTo(out, card);
    				printedSomething = true;
    			} else {
    				out.format(Locale.US, "%3s", "");
    			}
    		}
    		out.println();
    	} while (printedSomething);
    }
    
    @Override
    public int hashCode() {
        return 0xc284f7a1 ^ columns.hashCode() ^ Integer.rotateLeft(stockPile.hashCode(), 8)
                ^ Integer.rotateLeft(stockPileIndex, 16) ^ Integer.rotateLeft(foundation.hashCode(), 24);
    }
    
    @Override
    public boolean equals(final Object obj) {
        if (obj instanceof Board) {
            final Board board = (Board) obj;
            // Put stockPileIndex first since it changes a lot.
            if (stockPileIndex != board.stockPileIndex) {
            	return false;
            }
            if (stockPile.size() != board.stockPile.size()) {
                return false;
            }
            for (final Suit suit : Suit.values()) {
                if (foundation.get(suit).size() != board.foundation.get(suit).size()) {
                    return false;
                }
            }
            return columns.equals(board.columns)
            		&& foundation.equals(board.foundation)
                    && stockPile.equals(board.stockPile);
        } else {
            return false;
        }
    }
    
    @Override
    public String toString() {
        return "Board {columns: " + columns + ", stockPile: " + stockPile + ", stockPileIndex: " + stockPileIndex + ", foundation: " + foundation
                + "}";
    }
    
}

File: csharp/Solitaire/Klondike/Board.cs
using Silnith.Game.Deck;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Silnith.Game.Klondike
{
    /// <summary>
    /// The board for Klondike solitaire.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The board consists of seven columns, the foundation, and the stock pile.
    /// </para>
    /// </remarks>
    public class Board : IEquatable<Board?>
    {
        /// <summary>
        /// The columns of the board.  Each column consists of a number of face-down cards,
        /// and a run of face-up cards.
        /// </summary>
        public IReadOnlyList<Column> Columns
        {
            get;
        }

        /// <summary>
        /// The stock pile.
        /// </summary>
        public IReadOnlyList<Card> StockPile
        {
            get;
        }

        /// <summary>
        /// The current index into the stock pile.
        /// This is <c>0</c> if the stock pile has not been advanced,
        /// meaning no cards are available to be drawn.
        /// When this reaches the size of the stock pile, it is
        /// eligible to be recycled.
        /// </summary>
        public int StockPileIndex
        {
            get;
        }

        /// <summary>
        /// The foundation.
        /// When all cards are in the foundation, the game is won.
        /// </summary>
        public IReadOnlyDictionary<Suit, IReadOnlyList<Card>> Foundation
        {
            get;
        }

        /// <summary>
        /// Constructs a new board.  All parameters should be immutable.
        /// </summary>
        /// <remarks>
        /// <para>
        /// Parameters should be immutable.  This is not enforced in code.
        /// </para>
        /// </remarks>
        /// <param name="columns">The columns for the new board.</param>
        /// <param name="stockPile">The stock pile.</param>
        /// <param name="stockPileIndex">The index into the stock pile of the current draw card.
        /// <c>0</c> means no card is available to be drawn,
        /// <c>stockPile.Count</c> means all cards have been advanced
        /// and the last card is available to be drawn.</param>
        /// <param name="foundation">The foundation.</param>
        public Board(IReadOnlyList<Column> columns, IReadOnlyList<Card> stockPile, int stockPileIndex, IReadOnlyDictionary<Suit, IReadOnlyList<Card>> foundation)
        {
            if (stockPileIndex < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(stockPileIndex), "Stock pile index must be non-negative.");
            }
            if (stockPileIndex > stockPile.Count)
            {
                throw new ArgumentOutOfRangeException(nameof(stockPileIndex), "Stock pile index outside of stock pile.");
            }
            Columns = columns;
            StockPile = stockPile;
            StockPileIndex = stockPileIndex;
            Foundation = foundation;
        }

        /// <summary>
        /// Returns <see langword="true"/> if the stock pile can be advanced.
        /// If the stock pile is empty, or if the stock pile index
        /// is beyond the end of the stock pile, this is <see langword="false"/>.
        /// </summary>
        /// <remarks>
        /// <para>
        /// If the stock pile is empty, it cannot be advanced or recycled.
        /// Otherwise, either this or <see cref="CanRecycleStockPile"/>
        /// will be <see langword="true"/>, and the other <see langword="false"/>.
        /// </para>
        /// </remarks>
        /// <returns><see langword="true"/> if the stock pile can be advanced.</returns>
        public bool CanAdvanceStockPile()
        {
            return StockPileIndex < StockPile.Count;
        }

        /// <summary>
        /// Returns <see langword="true"/> if the stock pile can be recycled.
        /// If the stock pile is empty, or if the stock pile index
        /// is not currently beyond the end of the stock pile, this is <see langword="false"/>.
        /// </summary>
        /// <remarks>
        /// <para>
        /// If the stock pile is empty, it cannot be advanced or recycled.
        /// Otherwise, either this or <see cref="CanAdvanceStockPile"/>
        /// will be <see langword="true"/>, and the other <see langword="false"/>.
        /// </para>
        /// </remarks>
        /// <returns><see langword="true"/> if the stock pile can be recycled.</returns>
        public bool CanRecycleStockPile()
        {
            // 0 represents no cards flipped, so offset by one.
            return StockPileIndex > 0 && StockPileIndex >= StockPile.Count;
        }

        /// <summary>
        /// Returns the current card that can be drawn from the stock pile.
        /// </summary>
        /// <returns>The current card available from the stock pile.</returns>
        /// <exception cref="ArgumentOutOfRangeException">If no card is available to be drawn from the stock pile.</exception>
        public Card GetStockPileCard()
        {
            return StockPile[StockPileIndex - 1];
        }

        /// <summary>
        /// Returns the top card on the foundation for the given suit.
        /// </summary>
        /// <param name="suit">The suit.</param>
        /// <returns>The current top card for the suit in the foundation.</returns>
        /// <exception cref="ArgumentOutOfRangeException">If the foundation has no cards for <paramref name="suit"/>.</exception>
        public Card GetTopOfFoundation(Suit suit)
        {
            IReadOnlyList<Card> foundationForSuit = Foundation[suit];
            return foundationForSuit[foundationForSuit.Count - 1];
        }

        /// <summary>
        /// Returns a copy of the foundation with the given card added to it.
        /// </summary>
        /// <remarks>
        /// <para>
        /// This does no validation that the move is legal.
        /// </para>
        /// </remarks>
        /// <param name="card">The card to add to the foundation.</param>
        /// <returns>A copy of the foundation with one card added.</returns>
        public IReadOnlyDictionary<Suit, IReadOnlyList<Card>> GetFoundationPlusCard(Card card)
        {
            Suit suit = card.Suit;

            return new Dictionary<Suit, IReadOnlyList<Card>>(Foundation)
            {
                [suit] = new List<Card>(Foundation[suit])
                {
                    card,
                },
            };
        }

        /// <summary>
        /// Extracts the top card from the foundation for the given suit, and returns
        /// both the card and the remaining foundation missing the card.
        /// </summary>
        /// <param name="suit">The suit of the foundation to remove the top card from.</param>
        /// <returns>A tuple of the card, and the remaining foundation.</returns>
        public Tuple<Card, IReadOnlyDictionary<Suit, IReadOnlyList<Card>>> ExtractCardFromFoundation(Suit suit)
        {
            IReadOnlyList<Card> foundationForSuit = Foundation[suit];
            int suitCountMinusOne = foundationForSuit.Count - 1;
            Card card = foundationForSuit[suitCountMinusOne];
            IReadOnlyDictionary<Suit, IReadOnlyList<Card>> newFoundation = new Dictionary<Suit, IReadOnlyList<Card>>(Foundation)
            {
                [suit] = foundationForSuit.Take(suitCountMinusOne).ToList(),
            };
            return new Tuple<Card, IReadOnlyDictionary<Suit, IReadOnlyList<Card>>>(card, newFoundation);
        }

        /// <summary>
        /// Extracts the currently visible card from the stock pile, and returns both the card
        /// and the remaining stock pile missing the card.
        /// </summary>
        /// <returns>A tuple of the card, and the remaining stock pile.</returns>
        public Tuple<Card, IReadOnlyList<Card>> ExtractStockPileCard()
        {
            Card card = StockPile[StockPileIndex - 1];
            /*
             * StockPile: [a, b, c, d]
             * StockPileIndex: 2
             * 
             * Result: [a, c, d]
             */
            IReadOnlyList<Card> remainingStockPile = StockPile.Take(StockPileIndex - 1).Concat(StockPile.Skip(StockPileIndex)).ToList();
            return new Tuple<Card, IReadOnlyList<Card>>(card, remainingStockPile);
        }

        /// <summary>
        /// Checks and returns whether it would be legal to add the given card to the foundation.
        /// </summary>
        /// <param name="card">The card to compare against the foundation.</param>
        /// <returns><see langword="true"/> if it is legal to add the card to the foundation.</returns>
        public bool CanAddToFoundation(Card card)
        {
            return card.Value.GetValue() == 1 + Foundation[card.Suit].Count;
        }

        private void PrintCardTo(Card card)
        {
            Console.Write("{0,2}{1}", card.Value.ToSymbol(), card.Suit.ToSymbol());
        }

        public void PrintTo()
        {
            foreach (KeyValuePair<Suit, IReadOnlyList<Card>> keyValuePair in Foundation)
            {
                Console.Write("{0,2}", string.Empty);
                IReadOnlyList<Card> cards = keyValuePair.Value;
                if (cards.Any())
                {
                    PrintCardTo(cards[cards.Count - 1]);
                }
                else
                {
                    Console.Write("{0,3}", "--");
                }
            }
            Console.Write("{0,3}", string.Empty);
            Console.Write("({0,2:D}/{1,2:D})", StockPileIndex, StockPile.Count);
            Console.Write("{0,2}", string.Empty);
            if (StockPileIndex > 0)
            {
                PrintCardTo(StockPile[StockPileIndex - 1]);
            }
            else
            {
                Console.Write("{0,3}", string.Empty);
            }
            Console.WriteLine();
            List<IEnumerator<Card>> enumerators = new List<IEnumerator<Card>>(Columns.Count);
            foreach (Column column in Columns)
            {
                Console.Write("{0,2}", string.Empty);
                Console.Write("({0:D})", column.GetCountOfFaceDownCards());
                enumerators.Add(column.FaceUp.GetEnumerator());
            }
            Console.WriteLine();
            bool printedSomething;
            do
            {
                printedSomething = false;
                foreach (IEnumerator<Card> enumerator in enumerators)
                {
                    Console.Write("{0,2}", string.Empty);
                    if (enumerator.MoveNext())
                    {
                        PrintCardTo(enumerator.Current);
                        printedSomething = true;
                    }
                    else
                    {
                        Console.Write("{0,3}", string.Empty);
                    }
                }
                Console.WriteLine();
            } while (printedSomething);
        }

        /// <inheritdoc/>
        public override bool Equals(object? obj)
        {
            return Equals(obj as Board);
        }

        /// <inheritdoc/>
        public bool Equals(Board? other)
        {
            return other != null
                && Columns.SequenceEqual(other.Columns)
                && StockPile.SequenceEqual(other.StockPile)
                && StockPileIndex == other.StockPileIndex
                && Foundation[Suit.Club].SequenceEqual(other.Foundation[Suit.Club])
                && Foundation[Suit.Diamond].SequenceEqual(other.Foundation[Suit.Diamond])
                && Foundation[Suit.Heart].SequenceEqual(other.Foundation[Suit.Heart])
                && Foundation[Suit.Spade].SequenceEqual(other.Foundation[Suit.Spade]);
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            HashCode hashCode = new HashCode();
            foreach (Column column in Columns)
            {
                hashCode.Add(column);
            }
            foreach (Card card in StockPile)
            {
                hashCode.Add(card);
            }
            hashCode.Add(StockPileIndex);
            foreach (Card card in Foundation[Suit.Club])
            {
                hashCode.Add(card);
            }
            foreach (Card card in Foundation[Suit.Diamond])
            {
                hashCode.Add(card);
            }
            foreach (Card card in Foundation[Suit.Heart])
            {
                hashCode.Add(card);
            }
            foreach (Card card in Foundation[Suit.Spade])
            {
                hashCode.Add(card);
            }
            return hashCode.ToHashCode();
        }
    }
}
